### ISSUE TITLE
Bot Fails to Locate Elements on LinkedIn Job Search Page #70

### ISSUE THREAD

**Issue by jeinfra** [voidbydefault/EasyApplyBot#71]
The bot is failing to locate the necessary elements on LinkedIn's job search page, resulting in a NoSuchElementException. This issue seems to be caused by changes in LinkedIn's page structure or class names.

```
Attempting to restore previous session...
Starting the search for Systems engineer in Ireland.
Going to job page 0
Starting the application process for this page...
Traceback (most recent call last):
File "/Users/user/jobs/EasyApplyBot/linkedineasyapply.py", line 164, in apply_jobs
job_results = self.browser.find_element(By.CLASS_NAME, "jobs-search-results-list")
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/Users/user/jobs/EasyApplyBot/venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py", line 766, in find_element
return self.execute(Command.FIND_ELEMENT, {"using": by, "value": value})["value"]
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/Users/user/jobs/EasyApplyBot/venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py", line 380, in execute
self.error_handler.check_response(response)
File "/Users/user/jobs/EasyApplyBot/venv/lib/python3.12/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"css selector","selector":".jobs-search-results-list"}
(Session info: chrome=131.0.6778.86); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
Stacktrace:
0 chromedriver 0x0000000102f6bac4 cxxbridge1$str$ptr + 3651580
1 chromedriver 0x0000000102f64314 cxxbridge1$str$ptr + 3620940
2 chromedriver 0x00000001029cc4b4 cxxbridge1$string$len + 89224
3 chromedriver 0x0000000102a10898 cxxbridge1$string$len + 368748
4 chromedriver 0x0000000102a4a0fc cxxbridge1$string$len + 604368
5 chromedriver 0x0000000102a050b0 cxxbridge1$string$len + 321668
6 chromedriver 0x0000000102a05d00 cxxbridge1$string$len + 324820
7 chromedriver 0x0000000102f36e08 cxxbridge1$str$ptr + 3435328
8 chromedriver 0x0000000102f3a120 cxxbridge1$str$ptr + 3448408
9 chromedriver 0x0000000102f1e17c cxxbridge1$str$ptr + 3333812
10 chromedriver 0x0000000102f3a9e0 cxxbridge1$str$ptr + 3450648
11 chromedriver 0x0000000102f0f988 cxxbridge1$str$ptr + 3274432
12 chromedriver 0x0000000102f550f4 cxxbridge1$str$ptr + 3558956
13 chromedriver 0x0000000102f55270 cxxbridge1$str$ptr + 3559336
14 chromedriver 0x0000000102f63f88 cxxbridge1$str$ptr + 3620032
15 libsystem_pthread.dylib 0x000000018c0c32e4 _pthread_start + 136
16 libsystem_pthread.dylib 0x000000018c0be0fc thread_start + 8

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/Users/user/jobs/EasyApplyBot/linkedineasyapply.py", line 113, in start_applying
self.apply_jobs(location)
File "/Users/user/jobs/EasyApplyBot/linkedineasyapply.py", line 173, in apply_jobs
raise Exception("No more jobs on this page.")
Exception: No more jobs on this page.
Sleeping for 892.8907170295715 seconds.
^CTraceback (most recent call last):
File "/Users/user/jobs/EasyApplyBot/main.py", line 147, in
bot.start_applying()
File "/Users/user/jobs/EasyApplyBot/linkedineasyapply.py", line 133, in start_applying
time.sleep(time_left)
KeyboardInterrupt
```

**Comment by voidbydefault** [comment]
It's some recent change and Linkedin f**ked-up again. Issue is same as #66 , I will figure out a fix and post update.

**Comment by juda727** [comment]
Hello, So its gets the elements for the job list now however its outputs this error after:   job_list = self.browser.find_elements(By.CLASS_NAME, 'scaffold-layout__list-container')[0].find_elements(
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range. You getting this on your side as well

**Comment by juda727** [comment]
I'm thinking might have to XPATH that as well? what do you think?

**Comment by voidbydefault** [comment]
Index error is happening to me as well after some condition which I didn't test.

XPATH generally is not a good idea because structure of webpage can change more often than the web class elements plus XPATH is an absolute path which isn't the best way to refer elements (things are better if they're dynamic and modular).

Now the root cause is LinkedIn is using a seemingly random string to refer to `jobs-search-results-list`... I couldn't figure out a way to handle that part so the bot could resume scroll up or down functionality.

In short, XPATHing everything isn't the best idea plus it will have spiraling effects and we will end up changing everything one by one. Other option is to doable scrolling but it helps mimicking human behavior plus explores total jobs in search results page -- something I plan to test tomorrow. Alternatively, if you can explore and suggest a better fix then please do (but other than XPATH. Only reason I push a hot fix today because two other people also complained on my YT (if it's working no one gives a F).

**Comment by juda727** [comment]
Okay, thank you and we are all appreciative of your hard work to update this constantly. I will see if I can brainstorm some ways to bypass this as well.

**Comment by juda727** [comment]
Hello, could you paste the element page here, the section where we have to figure out a way for the bot to find the job list element. Thank you.

**Comment by voidbydefault** [comment]
Block in figure 1 used to be `job_results = self.browser.find_element(By.CLASS_NAME, "jobs-search-results-list")` but now it uses a random generated text string (see figure 2):

Figure 1
![image](https://github.com/user-attachments/assets/d3e03270-f12e-477c-bc7d-c7140056055e)

Figure 2
![image](https://github.com/user-attachments/assets/ad19a223-ed76-4a97-aa19-e5638d4734b1)

**Comment by lilyfaiz** [comment]
**Hi, I am getting ta similar error with the new patch.**  
Traceback (most recent call last):
  File "C:\Users\lilyr\EasyApplyBot\linkedineasyapply.py", line 189, in apply_jobs
    job_list = self.browser.find_elements(By.CLASS_NAME, 'scaffold-layout__list-container')[0].find_elements(
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\lilyr\EasyApplyBot\linkedineasyapply.py", line 115, in start_applying
    self.apply_jobs(location)
  File "C:\Users\lilyr\EasyApplyBot\linkedineasyapply.py", line 194, in apply_jobs
    raise Exception("No more jobs on this page.")
Exception: No more jobs on this page.
Sleeping for 803.2334249019623 seconds.

**Comment by voidbydefault** [comment]
> **Hi, I am getting ta similar error with the new patch.** Traceback (most recent call last): File 

Hi Lily, I am aware of this. It seems LI made a bunch of backend changes. I hope to resolve these over the weekend.

**Comment by lilyfaiz** [comment]
Hi, thank you so much for your contributions.  I am also trying to fix it (I've only so far located the problem) but I haven't gotten a solution yet.  The update to my code is still not recognizing the class name:
`div.display-flex.job-card-container.relative.job-card-list.job-card-container--clickable.job-card-list--underline-title-on-hover.jobs-search-results-list__list-item--active.jobs-search-two-pane__job-card-container--viewport-tracking-0"`

But I'm still working on it :) Hopefully you come to a fix before I do.

**Comment by voidbydefault** [comment]
I tried a few workarounds, but didn't go as expected. Bot trips to sleep unexpectedly. I am still trying to figure it out and hope to  get it fixed for all. 

Thanks for bearing with me.

**Comment by madingess** [comment]
Dear @voidbydefault , 

I sincerely appreciate the changes you've made to maintain this bot. I also wanted to thank you for updating me on your contributions to this project as LinkedIn pushed updated their front-end to ember.js and caused all these breaking changes. 

I've taken some time to review your changes and the outstanding issues. I merged many of your commits into the base repository https://github.com/madingess/EasyApplyBot. There I've also pushed some changes which correct the question fill up process to ensure questions can be answered again. The changes you may wish to cherry-pick into this repository are those from the pull request https://github.com/madingess/EasyApplyBot/pull/45. 

Please continue to inform me of any issues you may encounter with these changes or LinkedIn changes! 

I wish you the best, and happy holidays.

**Comment by voidbydefault** [comment]
Thanks mate @madingess . Merged your changes to my fork, additional questions working flawlessly as ever.  Closing issue.
